### PR TITLE
Reduce AST complexity: Literal

### PR DESCRIPTION
**Pull request by VladislavsPerkanuks**

### CODE REVIEW

File: parse/ast.go [hunk 1]
```diff
@@ -335,9 +330,10 @@ func (ae AnnotationExpression) validate() error {
 
 type QuotedLiteral string
 
-func (QuotedLiteral) node()    {}
-func (QuotedLiteral) literal() {}
-func (QuotedLiteral) value()   {}
+type NameLiteral string
+
+type NumberLiteral float64
```
Comment (klavinsjanis) [inline parse/ast.go:335]: shouldn't we support all numeric types?

Comment (klavinsjanis) [reply parse/ast.go:335]: Only allowed to pass in float64 as function argument?
What about supporting other types like int, unit, complex etc and approximations?

File: parse/ast.go [hunk 2]
```diff
@@ -347,6 +343,8 @@ func (ql QuotedLiteral) String() string {
 
 	return fmt.Sprintf("|%s|", r.Replace(string(ql)))
 }
+func (nl NameLiteral) String() string   { return string(nl) }
+func (nl NumberLiteral) String() string { return fmt.Sprint(float64(nl)) }
```
Comment (mvilks) [inline parse/ast.go:347]: Why do you need additional `float64()` cast?

File: parse/ast.go (hunk 1 above)
Comment (VladislavsPerkanuks) [reply parse/ast.go:335]: its not the scope of this PR, as it was not changed only moved.

but you can cast int, unit, exponent, to float64, with no problems.
The only problem is complex numbers

File: parse/ast.go (hunk 2 above)
Comment (VladislavsPerkanuks) [reply parse/ast.go:347]: without casting infinite recursion :)

File: parse/ast.go (hunk 1 above)
Comment (klavinsjanis) [reply parse/ast.go:335]: I guess add a comment :)

File: parse/ast.go (hunk 2 above)
Comment (VladislavsPerkanuks) [reply parse/ast.go:347]: as fmt.Sprint(nl), will call nl.String(), which will cal fmt.Sprint(nl), and so on and on

Comment (mvilks) [reply parse/ast.go:347]: Indeed.

File: parse/ast.go (hunk 1 above)
Comment (mvilks) [reply parse/ast.go:335]: complex numbers are not supported in MF2, the format is defined as JSON number https://www.rfc-editor.org/rfc/rfc8259#section-6

**Review by mvilks — APPROVED** [review]

File: parse/ast.go (hunk 1 above)
Comment (VladislavsPerkanuks) [reply parse/ast.go:335]: yeah seems correct, so all other types are safely can be converted to float64, so no problemo

**Review by klavinsjanis — APPROVED** [review]

**Pull request merged**